calculate.go

```diff
@@ -120,15 +130,23 @@ func (o *OATH) calculateAll() (map[string]string, error) {
 
 		switch tv.tag {
 
-		case 0x71:
+		case TAG_NAME:
 			names = append(names, string(tv.value))
 
-		case 0x7c:
+		case TAG_TOUCH:
 			codes = append(codes, touchRequired)
 
-		case 0x76:
+		case TAG_RESPONSE:
+			o.Debug("tag no full response: %x", tv.value)
+			return nil, fmt.Errorf("unable to handle full response %x", tv.value)
+
+		case TAG_TRUNCATED_RESPONSE:
 			codes = append(codes, otp(tv.value))
 
+		case TAG_NO_RESPONSE:
+			o.Debug("tag no response. Is HOTP %x", tv.value)
+			return nil, fmt.Errorf("unable to handle HOTP response %x", tv.value)
```

Comment:
Errors messages should generally be defined as private constants on the top of the file, prefixed with `err`. The actual error should be wrapped using the `errors` package like so: `errors.Wrapf(err, const-message, parameters...)` 

This patterns makes it easy to see what planned errors could be yielded by package.

Regarding the `0x77` - under which circumstances does this appear?

Reply:
Sorry. I think I understood what you meant here, but now I'm a little confused.

It looks like all the error strings you used are at the top of `ykoath.go`, but only as strings. So one can't use `errors.Is` to determine if they are a particular error.

Would it make sense to have those constants actually be `error`s? Eg. `const errFailedToListSuitableReader = errors.New("failed to connect to reader")`.

This, however doesn't allow formatting as is done with `errUnknownTag`. The go docs seem to suggest creating a struct that contains a field for the `tag` and then defining an `Error() string` function on that struct that returns the formatted text, however that's a bit convoluted. The other thing I can think of is defining `const errUnknownTag = errors.New("unknown tag")` and then catching and returning `fmt.Errorf("%w (%s)", errUnknownTag, tag)` so that one handling the error could still tell that it wraps `errUnknownTag`.

I'm really a go hobbyist, so I'm not as familiar with the current best practices for errors in go as they seem to change every couple of months. 😄